experimental/web_dashboard/lib/src/widgets/categories_dropdown.dart

```diff
@@ -24,61 +24,67 @@ class CategoryDropdown extends StatefulWidget {
 
 class _CategoryDropdownState extends State<CategoryDropdown> {
   Category _selected;
-  List<Category> _categories = [];
-  StreamSubscription _subscription;
 
-  void initState() {
-    super.initState();
-    _fetch();
-    _subscription = widget.api.subscribe().listen((categories) {
-      setState(() {
-        _categories = categories;
+  @override
+  Widget build(BuildContext context) {
+    return FutureBuilder<List<Category>>(
+      future: widget.api.list(),
+      builder: (context, futureSnapshot) {
+        // Show an empty dropdown while the data is loading
+        if (!futureSnapshot.hasData) {
+          return DropdownButton<Category>(items: [], onChanged: null);
+        }
+
+        // Select the first item in the dropdown if this is the first snapshot.
+        if (_selected == null && futureSnapshot.data.isNotEmpty) {
 
-        // Update the selected value if it doesn't exist in the new list.
-        if (categories.isNotEmpty && !categories.contains(_selected)) {
-          _selected = _categories.first;
-        } else if (categories.isEmpty) {
-          _selected = null;
+            _setSelected(futureSnapshot.data.first);
         }
-      });
-    });
-  }
 
-  void dispose() {
-    _subscription.cancel();
-    super.dispose();
-  }
+        return StreamBuilder<List<Category>>(
+          initialData: futureSnapshot.hasData ? futureSnapshot.data : [],
+          stream: widget.api.subscribe(),
+          builder: (context, snapshot) {
+            var data = snapshot.hasData ? snapshot.data : <Category>[];
 
-  Future _fetch() async {
-    var categories = await widget.api.list();
+            // Select the first element if the new list doesn't contain the
+            // selected value,
+            if (!snapshot.data.contains(_selected) &&
+                snapshot.data.isNotEmpty) {
+                _setSelected(snapshot.data.first);
+            }
+            return DropdownButton<Category>(
+              value: _selected,
+              items: data.map(_buildDropdownItem).toList(),
+              onChanged: (category) {
+                setState(() {
+                  _setSelected(category);
+                });
+              },
+            );
+          },
+        );
+      },
+    );
+  }
 
-    if (categories.isEmpty) {
+  void _setSelected(Category category) {
+    if (_selected == category) {
       return;
     }
+    _selected = category;
 
-    setState(() {
-      _categories = categories;
+    // Avoid invoking this callback during a build() because 1. calling setState
+    // during a build is an error, and 2. any widgets listening could be calling
+    // setState(). This can happen if the data is immediately returned by the
+    // FutureBuilder or StreamBuilder.
+    Future(() {
```

Comment:
This is an interesting problem. The app needs the list of categories to set the value of _selected, but it also needs the Future that's providing them to use in the FutureBuilder.

One thing you might do is try to move that assignment into a different place. I believe you could:

* Add an override for initState that
  - invokes widget.api.list()
  - stores the Future in a field in the State object
  - uses `then` to add a clause calling setState, assigning _selected, and returning the same list of categories.
* That stored Future could then be given to the FutureBuilder.
* The builder methods for the FutureBuilder/StreamBuilder could check the value of _selected, and (if it's null) just display an empty list as they do now if there's no data yet.

It's also possible that you could pull this off with a single call to `subscribe` (assuming it's a broadcast stream), by getting a reference to the stream in initState, assigning the stream to a property in the state object for use in `build`, and then using `.first.then` on the stream to call setState once you've gotten at least one value.

This is all speculative since I haven't tried it (and probably too large a suggestion), so I'm just throwing it out here while also LGTMing the PR. 😄

Reply:
This is a good suggestion, here's how this ended up: https://github.com/flutter/samples/pull/421/commits/34796cec49b54f8cfd460e71ac7dc1c04cbf4a3e